index.d.ts

```diff
@@ -0,0 +1,456 @@
+declare module 'oidc-provider' {
+    import { Request, Response, Handler } from 'express';
+    import Koa = require('koa');
+    import Router = require('koa-router');
+    import { IncomingMessage, ServerResponse } from 'http';
+    import EventEmitter = require('events');
+    import { JWK } from 'node-jose';
+
+    type SubjectType = 'public' | 'parwise';
+
+    interface IClient {
+        applicationType?: 'web' | 'native';
+        grant_types?: string[];
+        id_token_signed_response_alg?: string;
+        require_auth_time?: boolean;
+        response_types?: string[];
+        subject_type?: SubjectType;
+        token_endpoint_auth_method?: string;
+        request_uris?: string[];
+        client_id?: string;
+        client_secret?: string;
+        redirect_uris: string[];
+        introspection_endpoint_auth_method?: string;
+    }
+
+    interface IAdapter {
+        upsert: (id: string, payload: string, expiresIn: number) => Promise<void>;
+        find: (id: string) => Promise<void>;
+        findByUserCode: (userCode: string) => Promise<void>;
+        consume: (id: string) => Promise<void>;
+        destroy: (id: string) => Promise<void>;
+        connect?: (provider: Provider) => Promise<void>;
+        key: (id: string) => string;
+    }
+
+    interface IKeystore {
```

Comment:
IKeystore is unused, but i'm wondering if it's gonna be better to maintain own JWK.KeyStore types here.

Reply:
I'm thinking that would make sense. Could you give me an example structure?

Reply:
I'm thinking the types can just specify a jwks object as an interface, e.g. 

```js
{
  keys: [
    rsaKey | ecKey
    , ...
  ]
};

rsaKey = {
  kty: 'RSA',
  d: required string
  dp: required string
  dq: required string
  e: required string
  n: required string
  p: required string
  q: required string
  qi: required string
  use: optional string 'sig' | 'enc'
  kid: optional string,
  alg: optional string,
}

ecKey = {
  kty: 'EC',
  crv: required string
  x: required string
  y: required string
  d: required string
  use: optional string 'sig' | 'enc'
  kid: optional string,
  alg: optional string,
}
```

i believe most people use it this way anyway, passing the jose jwk keystore is just a QoL change.